source/common/common/callback_impl.cc

```diff
@@ -0,0 +1,64 @@
+#include "source/common/common/callback_impl.h"
+
+namespace Envoy {
+namespace Common {
+
+ThreadSafeCallbackHandlePtr ThreadSafeCallbackManager::add(Event::Dispatcher& dispatcher,
+                                                           Callback callback) {
+  Thread::LockGuard lock(lock_);
+  auto new_callback = std::make_shared<CallbackHolder>(*this, callback);
+  callbacks_.push_back(CallbackListEntry(std::weak_ptr<CallbackHolder>(new_callback), dispatcher));
+  // Get the list iterator of added callback handle, which will be used to remove itself from
+  // callbacks_ list.
+  new_callback->it_ = (--callbacks_.end());
+  return new_callback;
+}
+
+void ThreadSafeCallbackManager::runCallbacks() {
+  Thread::LockGuard lock(lock_);
+  for (auto it = callbacks_.cbegin(); it != callbacks_.cend();) {
+    auto& [cb, cb_dispatcher] = *(it++);
+
+    // sanity check cb is valid before attempting to schedule a dispatch
+    if (cb.expired()) {
+      continue;
+    }
+
+    cb_dispatcher.post([cb = cb] {
+      // Once we're running on the thread that scheduled the callback, validate the
+      // callback is still valid and execute
+      std::shared_ptr<CallbackHolder> cb_shared = cb.lock();
+      if (cb_shared != nullptr) {
+        cb_shared->cb_();
+      }
+    });
+  }
+}
+
+size_t ThreadSafeCallbackManager::size() const noexcept {
+  Thread::LockGuard lock(lock_);
+  return callbacks_.size();
+}
+
+void ThreadSafeCallbackManager::remove(typename std::list<CallbackListEntry>::iterator& it) {
+  Thread::LockGuard lock(lock_);
+  callbacks_.erase(it);
+}
+
+ThreadSafeCallbackManager::CallbackHolder::CallbackHolder(ThreadSafeCallbackManager& parent,
+                                                          Callback cb)
+    : cb_(cb), parent_dispatcher_(parent.dispatcher_), parent_(parent),
+      still_alive_(parent.still_alive_) {}
+
+ThreadSafeCallbackManager::CallbackHolder::~CallbackHolder() {
+  parent_dispatcher_.post([still_alive = still_alive_, &parent = parent_, it = it_]() mutable {
+    // We're running on the same thread the parent is managed on. We can assume there is
+    // no race between checking if alive and calling remove.
+    if (!still_alive.expired()) {
+      parent.remove(it);
+    }
+  });
+}
```

Comment:
Previous discussion on implementation [here](https://github.com/envoyproxy/envoy/pull/16201/files/a1846b1d555d12f0c6df646306b3c64b4550b0ae#r635665730).

----
> Could you say more about what happened here? Could you call parent.remove(it) directly instead of going through the dispatcher?
> 
> I was going to comment on still_alive_ below but realized that calling parent.remove based on still_alive_ doesn't work because the parent could be deleted between checking for still_alive_ and the actual call to `parent.remove()`. It may be possible to make this work out by keeping a weak_ptr to the parent, so it can be locked when attempting a removal. Another option to consider would be to ASSERT that all callback holders have been deleted prior to deletion of the parent.

We could call `parent.remove()` if we take a `weak_ptr` to the parent. The trade-off (which I'll let you weigh in on) would be that all instances of `ThreadSafeCallbackManager` would need to be through a `shared_ptr` and then we could make use of `std::enable_shared_from_this` to enable `weak_from_this`. 

I'm not against taking this approach. The main reason I took the approach I did was because (at least to me), I was able to reason about it more easily. Using the dispatcher means I'm running on the thread that manages the lifetime of the callback-manager and I can use the `still_alive_` ptr to validate the parent and there are not visible effects outside of the class (such as requiring the use of a `shared_ptr`).

if there are strong opinions on style or other considerations I'm not taking into account, I'm happy to change the approach here.

Reply:
Ah, one point that I didn't address specifically
> ... realized that calling parent.remove based on still_alive_ doesn't work because the parent could be deleted between checking for still_alive_ and the actual call to parent.remove(). 

This should be okay under the assumption that the parent dispatcher this is executing on is the same thread in which the lifetime of the callback-manager is managed. Meaning that there shouldn't be any process that frees the parent between the check to is_alive and the `parent.remove()` call.

The caveat is that this is an assumption based on the documented, supported use-cases of the class and not guaranteed to be correct. The approach with the shared/weak ptrs would be guaranteed.

Reply:
I've thought about this a bit and I think I agree that we should opt for the solution that provides more correctness guarantees given that a resulting bug could be difficult to troubleshoot. I've refactored the following:

0. The `ThreadSafeCallbackManager` must now be created via `create` (static method) which enforces the creation via a `shared_ptr`. I'm enforcing this at construction since it is now necessary to it's internal operation.
1. The parent is passed to the `CallbackHolder` via `weak_from_this`
2. The callback's dispatcher is used to assert destruction of `CallbackHolder` on the thread that the callback was meant to be executed on.
3. A lock is taken directly on the `parent` via `remove()` and no use of an async `Dispatcher::post()` is used